hilti/toolchain/src/compiler/global-optimizer.cc

```diff
@@ -0,0 +1,235 @@
+// Copyright (c) 2020-2021 by the Zeek Project. See LICENSE for details.
+
+#include "hilti/compiler/global-optimizer.h"
+
+#include <optional>
+#include <unordered_set>
+#include <utility>
+
+#include <hilti/ast/ctors/default.h>
+#include <hilti/ast/declarations/function.h>
+#include <hilti/ast/detail/visitor.h>
+#include <hilti/ast/expressions/ctor.h>
+#include <hilti/ast/node.h>
+#include <hilti/ast/types/struct.h>
+#include <hilti/base/logger.h>
+#include <hilti/base/timing.h>
+#include <hilti/base/util.h>
+
+namespace hilti {
+
+namespace logging::debug {
+inline const DebugStream GlobalOptimizer("global-optimizer");
+}
+
+enum class Stage { COLLECT, PRUNE };
+
+template<typename T>
+std::optional<std::pair<ID, ID>> typeID(T&& x) {
+    auto id = x.typeID();
+    if ( ! id )
+        return {};
+
+    auto id_components = util::split(*id, "::");
+    if ( id_components.size() < 2 )
+        return {};
+
+    auto unit_id = *id_components.rbegin();
+    auto module_id = *++id_components.rbegin();
+
+    return {std::pair<ID, ID>{module_id, unit_id}};
+}
+
+struct Visitor : hilti::visitor::PreOrder<bool, Visitor> {
+    Visitor(GlobalOptimizer::Hooks* hooks) : _hooks(hooks) {}
+
+    static void removeNode(position_t& p) { p.node = node::none; }
+
+    Stage _stage = Stage::COLLECT;
+
+    void collect(Node& node) {
+        _stage = Stage::COLLECT;
+
+        for ( auto i : this->walk(&node) )
+            dispatch(i);
+    }
+
+    void prune(Node& node) {
+        _stage = Stage::PRUNE;
+
+        bool modified = false;
+
+        for ( auto&& [hook_id, uses] : *_hooks ) {
+            // Linker joins are implemented via functions, so if we remove all
+            // functions data dependencies (e.g., needed for subunits) might
+            // get broken. Leave at least one function in unit so it gets emitted.
+            //
+            // TODO(bbannier): Explicitly express data dependencies in joins.
+            if ( std::get<2>(hook_id) == std::string("__str__") ) {
+                uses.set(1);
+                continue;
+            }
+        }
+
+        for ( auto i : this->walk(&node) ) {
+            if ( auto x = dispatch(i) )
+                modified = modified && x;
+        }
+
+        if ( ! modified )
+            return;
+
+        prune(node);
+    }
+
+    result_t operator()(const type::struct_::Field& x, position_t p) {
+        if ( auto type_ = x.type().tryAs<type::Function>(); ! type_ )
+            return false;
+
+        auto fn = x.childsOfType<Function>();
+        assert(fn.size() <= 1);
+
+        bool is_keep = ! fn.empty() && AttributeSet::find(fn.front().attributes(), "&keep");
```

Comment:
any idea for a more self-explanatory attribute name? `&no-optimize` perhaps?

Reply:
This isn't really about _not optimizing_, but more about _keep a function which is part of some API_, which I think `&no-optimize` doesn't capture (and we might even end up optimizing the internals of such functions in the future).

WDYT about e.g., `&always_emit` or `&external_api`?

Reply:
I like `&always-emit` , we already `&no-emit` elsewhere, that matches.